File: 82. Remove Duplicates from Sorted List II/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
    var prev *ListNode
    cur := head
    
    // Remove all the repeats except for the first occurance and remember the repeated numbers
    repeats := make(map[int]bool)
    for cur != nil {
        if prev != nil && prev.Val == cur.Val {
            repeats[cur.Val] = true
            prev.Next = cur.Next
        } else {
            repeats[cur.Val] = false
            prev = cur  
        }
        cur = cur.Next
    }
    
    // Remove all the single occurence of repeated numbers
    for head != nil && repeats[head.Val] == true {
        head = head.Next
    }
    if head == nil {
        return head
    }
    prev = head
    cur = head.Next
    for cur != nil {
        if repeats[cur.Val] == true {
            prev.Next = cur.Next
        } else {
            prev = cur
        }
        cur = cur.Next
    }
    return head
}

File: 82. Remove Duplicates from Sorted List II/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: 'ListNode') -> 'ListNode':
        seen = set()
        remove = set()
        node = ListNode(None)
        node.next = head
        head = node
        node = node.next
        
        while node:
            if not node.val in remove:                
                if node.val in seen:
                    seen.remove(node.val)
                    remove.add(node.val)
                else:
                    seen.add(node.val)

            node = node.next
            
        prev = head
        node = head.next
        while node:
            if node.val in remove:
                prev.next = node.next
            else:
                prev = node
                
            node = node.next
            
        return head.next
